Implement let-else type annotations natively

cormacrelf opened the pull request:
Tracking issue: #87335

Fixes #89688, fixes #89807, edit: fixes  #89960 as well

As explained in https://github.com/rust-lang/rust/issues/89688#issuecomment-940405082, the previous desugaring moved the let-else scrutinee into a dummy variable, which meant if you wanted to refer to it again in the else block, it had moved.

This introduces a new hir type, ~~`hir::LetExpr`~~ `hir::Let`, which takes over all the fields of `hir::ExprKind::Let(...)` and adds an optional type annotation. The `hir::Let` is then treated like a `hir::Local` when type checking a function body, specifically:

* `GatherLocalsVisitor` overrides a new `Visitor::visit_let_expr` and does pretty much exactly what it does for `visit_local`, assigning a local type to the `hir::Let` ~~(they could be deduplicated but they are right next to each other, so at least we know they're the same)~~
* It reuses the code in `check_decl_local` to typecheck the `hir::Let`, simply returning 'bool' for the expression type after doing that.

* ~~`FnCtxt::check_expr_let` passes this local type in to `demand_scrutinee_type`, and then imitates check_decl_local's pattern checking~~
* ~~`demand_scrutinee_type` (the blindest change for me, please give this extra scrutiny) uses this local type instead of of creating a new one~~
    * ~~Just realised the `check_expr_with_needs` was passing NoExpectation further down, need to pass the type there too. And apparently this Expectation API already exists.~~

Some other misc notes:

* ~~Is the clippy code supposed to be autoformatted? I tried not to give huge diffs but maybe some rustfmt changes simply haven't hit it yet.~~
* in `rustc_ast_lowering/src/block.rs`, I noticed some existing `self.alias_attrs()` calls in `LoweringContext::lower_stmts` seem to be copying attributes from the lowered locals/etc to the statements. Is that right? I'm new at this, I don't know.

cormacrelf commented:
Ok, we're back in business I think. Let-else is now type-checked exactly as a let statement is checked; they use the same code path via a `Declaration` abstraction. Obviously this raises a fair question, why not just use `hir::Local` all the way? I thought of this, tried it, and it was painful.

- The `init` being optional on a Local means lots of `local.init.unwrap_or_else(|| span_bug!(local.span, "..."))`. It's nice to have a type that maps exactly to the construct, and this remains true for future use cases e.g. `Guard::IfLet`.
- The visitor implementations got knotty, wondering whether you need to `.visit_local(local)` or just use its component parts. It's less confusing this way, at least for me, it forces you to think about them differently and I think that is generally justified by the non-optional inititaliser.

I have added a bunch of tests, one adapted from @est31's rustc work (see the tracking issue) and a bunch adapted from rfc2005 match ergonomics to check that that all works, especially with type annotations attached.

bors commented:
:umbrella: The latest upstream changes (presumably #90000) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #90126) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nagisa commented on compiler/rustc_typeck/src/check/gather_locals.rs:
```diff
@@ -7,6 +7,31 @@ use rustc_middle::ty::Ty;
+    pub ty: Option<&'a hir::Ty<'a>>,
+    pub span: Span,
+    pub init: Option<&'a hir::Expr<'a>>,
+}
```
Addition of this further suggests to me that the HIR representation of `let-else` should have been modeled after a `hir::Local` and not `if-let`… Out of scope for this PR though.

cormacrelf replied:
At least we could pretty easily add support for type annotations to `if-let`, i.e. this which you can't do today:

```rust
if let Some(x): &mut i32 = optional() { ... }
```

If anything, of all the name bindings in the language, `hir::Local` is the odd one out, having an optional init expression.

cormacrelf replied:
I guess a more perfect refactoring is to extract the common parts (excluding the ty/option ty) into a `hir::Decl` or something. But similarly this is a big enough change already without doing to `hir::Local` what I've done to let-else.

Ultimately it's just not that big a deal for the codebase. The other thing that's apparent from `gather_locals.rs` is that we only needed common functionality in one place and the rest of a change like that would just be busywork.

nagisa commented on compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:
```diff
@@ -890,19 +896,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     }
 
-    fn overwrite_local_ty_if_err(
+    pub(crate) fn overwrite_local_ty_if_err(
```
This doesn't need to become `pub(crate)` does it?

nagisa reviewed (commented):
One more test case that would be nice is for `ref` bindings (i.e. references introduced to bindings manually rather than through the match ergonomics stuff). There is one test with `ref mut` but given how thorough the autoderef ones are, the `ref mut` one looks pretty underrepresented.

See the comments inline, but the overall feeling I have is that the compiler side of the changes are pretty good.

nagisa commented:
r? @nagisa

JohnCSimon commented:
@cormacrelf ping from triage
Can you please address the review comments and then set this PR to be ready for review?

cormacrelf commented on src/test/ui/let-else/issue-89960.rs:
```diff
@@ -1,7 +1,9 @@
+// This had an ICE, see issue #89960
+
 #![feature(let_else)]
+#![deny(unused_variables)]
```
This is now covered by `let-else-ref-bindings-pass.rs`

cormacrelf commented on src/test/ui/let-else/let-else-binding-explicit-mut-annotated.rs:
```diff
@@ -0,0 +1,20 @@
+    *n += 1;
+    let _ = n;
+
+    let Some(n): &mut Option<i32> = &mut &mut Some(5i32) else { return }; // OK
```
I've moved the passing cases into `let-else-binding-explicit-mut-pass.rs`

cormacrelf commented on src/test/ui/let-else/let-else-deref-coercion.rs:
```diff
@@ -0,0 +1,90 @@
+}
+impl Foo {
+    fn set_bar(&mut self, value: u32) {
+        let Bar::Present(z): &mut Bar = self else { // OK
```
I've split this file into two similar ones, one `check-pass` = `let-else-deref-coercion-annotated.rs`

cormacrelf commented:
Alright, changes applied as requested. I've rebased this and fixed the merge conflicts with the recent clippy changes.

The only interesting thing from the rebase is that the "try `self.0`" suggestion no longer appears, and thinking about that briefly, you realise the suggestion was wrong (you would have to write `&mut self.0`, so someone's rightly disabled this.

nagisa commented on compiler/rustc_typeck/src/check/expr.rs:
```diff
@@ -119,7 +119,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     }
 
-    fn check_expr_with_expectation_and_needs(
+    pub(super) fn check_expr_with_expectation_and_needs(
```
Nit: Does this need to become a `pub(super)`?

nagisa commented:
Can you please also squash the history slightly so that it conveys meaningful information once it is merged into the mainline?

r=me after! Thanks for taking time to do all this!

cormacrelf commented:
@nagisa i think that's done, thanks for your help!

nagisa commented:
@bors r+ rollup=never

Thanks!

bors commented:
:pushpin: Commit fec8a507a27de1b08a0b95592dc8ec93bf0a321a has been approved by `nagisa`

<!-- @bors r=nagisa fec8a507a27de1b08a0b95592dc8ec93bf0a321a -->
<!-- homu: {"type":"Approved","sha":"fec8a507a27de1b08a0b95592dc8ec93bf0a321a","approver":"nagisa"} -->

bors commented:
:hourglass: Testing commit fec8a507a27de1b08a0b95592dc8ec93bf0a321a with merge dde825db464b08d6f572766579dfb629b837368c...
<!-- homu: {"type":"BuildStarted","head_sha":"fec8a507a27de1b08a0b95592dc8ec93bf0a321a","merge_sha":"dde825db464b08d6f572766579dfb629b837368c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4567010697?check_suite_focus=true)
Approved by: nagisa
Pushing dde825db464b08d6f572766579dfb629b837368c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4567010697?check_suite_focus=true"},"merge_sha":"dde825db464b08d6f572766579dfb629b837368c"} -->

rust-timer commented:
Finished benchmarking commit (dde825db464b08d6f572766579dfb629b837368c): [comparison url](https://perf.rust-lang.org/compare.html?start=7abab1efb21617ba6845fa86328dffa16cfcf1dc&end=dde825db464b08d6f572766579dfb629b837368c).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
